### ISSUE TITLE
Rectangle doesn't rerender when Stroke's DynamicResource SolidColorBrush changes

### ISSUE THREAD

**Issue by SKProCH** [AvaloniaUI/Avalonia#12408]
**Describe the bug**
When i change color of SolidColorBrush, which bind with DynamicResource to Rectangle's Stroke, then Rectangle doesn't rerender.

**To Reproduce**
Repro project: [AvaloniaApplication1.zip](https://github.com/AvaloniaUI/Avalonia/files/12229165/AvaloniaApplication1.zip)

1. Add a SolidColorBrush to resources (for example in App)
    ```xaml
    <Application.Resources>
      <SolidColorBrush x:Key="TestBrush" Color="Red" />
    </Application.Resources>
    ```
2. Add a rect, set Stroke property
    ```xaml
    <Rectangle Stroke="{StaticResource TestBrush}" StrokeThickness="1" Height="10"></Rectangle>
    ```
3. Add a way to change TestBrush color (button callback in my case)
    ```cs
    private void Button_OnClick(object? sender, RoutedEventArgs e) {
        var solidColorBrush = App.Current.Resources.First().Value as SolidColorBrush;
        var random = new Random();
        solidColorBrush.Color = Color.FromRgb((byte)random.Next(0, 256), (byte)random.Next(0, 256), (byte)random.Next(0, 256));
    }
    ```
4. Launch app. click button, see nothing changed.

**Expected behavior**
Rectangle should rerender when Stroke SolidColorBrush Color changed

**Screenshots**

https://github.com/AvaloniaUI/Avalonia/assets/29896317/b5e88a89-af91-4e60-9d3a-ad7a0c3b150a

**Desktop (please complete the following information):**

- Version: 11.0.1

**Additional context**
Rectangle.Fill property working fine.
Also related to https://github.com/AvaloniaUI/Avalonia/issues/11673

**Comment by SKProCH** [comment]
Additional info: when we have different ThemeDictionaries with target brush, switching between them works fine for Rectangle.Stroke. 

But issue still valid and plenty of people searched for fix in Material.Avalonia.

**Comment by SKProCH** [comment]
Seems like this is affects `Border`'s BorderBrush as well. @kekekeks there is no other options to property draw borders with changeable SolidColorBrush.

Original issue about `Border` reported here: https://github.com/AvaloniaCommunity/Material.Avalonia/issues/325

Also, I've updated an repro and put it into a git repository: https://github.com/SKProCH/AvaloniaBugsRepros (`BorderBindingToSolidColorBrush` project) to provide an easy way to reproduce it.

https://github.com/AvaloniaUI/Avalonia/assets/29896317/2c5cf554-9d87-474f-b649-a2fe600a941a

**Comment by SKProCH** [comment]
@kekekeks seems like i found a root cause of this problem. 

As far as I know, Render method now only define how the control should be rendered by composition renderer, since it called once and other brushes updates do not call the Render method (but control actually re-rendered when `Fill` brush changes by the compositor). 

If we look at the code for rendering Shape:   

https://github.com/AvaloniaUI/Avalonia/blob/63a5bcce4cb990277cbfea0908c808251b502a25/src/Avalonia.Controls/Shapes/Shape.cs#L217-L225

We pass the `Fill` brush directly to `DrawGeometry`, but for the `Stroke` property we calls `.ToImmutable()`. I think what this is the reason why composition renderer property re-render control when `Fill` brush value changed, but do not properly re-render control when `Stroke` brush value changes - cuz we pass the immutable value to DrawingContext.

So, if we change it to:
```csharp
    pen = new Pen(
        stroke,
        StrokeThickness,
        dashStyle,
        StrokeLineCap,
        StrokeJoin);
}
context.DrawGeometry(Fill, pen, geometry);
```
everything start works properly.

**Comment by SKProCH** [comment]
Fixed in master branch - `11.1.999-cibuild0043160-beta`
